datumaro/components/project.py

```diff
@@ -2,486 +2,2383 @@
+                    to_visit.extend(uninitialized_parents)
+                    continue
+
+            type_ = BuildStageType[current['config'].type]
+            params = current['config'].params
+            if type_ == BuildStageType.transform:
+                kind = current['config'].kind
+                try:
+                    transform = self._tree.env.transforms[kind]
+                except KeyError as e:
+                    raise UnknownStageError("Unknown transform '%s'" % kind) \
+                        from e
+
+                dataset = _join_parent_datasets()
+                dataset = dataset.transform(transform, **params)
+
+            elif type_ == BuildStageType.filter:
+                dataset = _join_parent_datasets()
+                dataset = dataset.filter(**params)
+
+            elif type_ == BuildStageType.inference:
+                kind = current['config'].kind
+                model = self._project.make_model(kind)
+
+                dataset = _join_parent_datasets()
+                dataset = dataset.run_model(model)
+
+            elif type_ == BuildStageType.source:
+                assert len(initialized_parents) == 0, current_name
+
+                # The only valid situation we get here is that it is a
+                # generated source:
+                # - No cache entry
+                # - No local dir data
+                source_name = ProjectBuildTargets.strip_target_name(current_name)
+                source = self._tree.sources[source_name]
+                if not source.is_generated:
+                    raise MissingObjectError("Can't initialize stage '%s': "
+                        "source data is not available" % current_name)
+
+                if self._tree.is_working_tree:
+                    source_dir = self._project.source_data_dir(source_name)
+                else:
+                    source_dir = None
+                dataset = ProjectSourceDataset(source_dir, self._tree,
+                    source_name, readonly=not source_dir)
 
-                subsets[item.subset].items[item.id] = item
+            elif type_ == BuildStageType.project:
+                dataset = _join_parent_datasets(force=True)
 
-        # override with our items, fallback to existing images
-        if own_source is not None:
-            log.debug("Loading own dataset...")
-            for item in own_source:
-                existing_item = subsets.setdefault(
-                        item.subset, self.Subset(self, item.subset)). \
-                    items.get(item.id)
-                if existing_item is not None:
-                    item = item.wrap(path=None,
-                        image=ExactMerge.merge_images(existing_item, item))
+            elif type_ == BuildStageType.convert:
+                dataset = _join_parent_datasets()
 
-                subsets[item.subset].items[item.id] = item
+            else:
+                raise UnknownStageError("Unknown stage type '%s'" % type_)
```

Comment:
It is not used yet, but I see it so that users can supply their own pipeline(-like?) file in `datum apply` command, which executes a bunch of operations instead of a single one like `datum transform` does. So it is more of pipeline validation code, and the message could be `unexpected stage ...`. OTOH, it can also be seen as a regular safety fallback for missing cases of a switch-case block, where even `assert False` could fit. Probably, the `unexpected` one looks better.